shell/platform/linux/public/flutter_linux/fl_method_channel.h

```diff
@@ -0,0 +1,190 @@
+ * @name: a channel name
+ * @codec: the method codec
+ *
+ * Creates a new method channel. @codec must match the codec used on the Dart
+ * end of the channel.
+ *
+ * Returns: a new #FlMethodChannel.
+ */
+FlMethodChannel* fl_method_channel_new(FlBinaryMessenger* messenger,
+                                       const gchar* name,
+                                       FlMethodCodec* codec);
+
+/**
+ * fl_method_channel_set_method_call_handler:
+ * @channel: a #FlMethodChannel
+ * @handler: function to call when a method call is received on this channel
+ * @user_data: (closure): user data to pass to @handler
+ *
+ * Sets the function called when a method is called. Call
+ * fl_method_channel_respond_success(), fl_method_channel_respond_error() or
+ * fl_method_channel_respond_not_implemented() when the method completes.
+ * Ownership of #FlMethodChannelResponseHandle is transferred to the caller, and
+ * the call must be responded to to avoid memory leaks.
+ */
+void fl_method_channel_set_method_call_handler(
+    FlMethodChannel* channel,
+    FlMethodChannelMethodCallHandler handler,
+    gpointer user_data);
+
+/**
+ * fl_method_channel_invoke_method:
+ * @channel: a #FlMethodChannel
+ * @method: the method to call
+ * @args: (allow-none): arguments to the method, must match what the
+ * #FlMethodCodec supports
+ * @cancellable: (allow-none): a #GCancellable or %NULL
+ * @callback: (scope async): (allow-none): a #GAsyncReadyCallback to call when
+ * the request is satisfied or %NULL to ignore the response
+ * @user_data: (closure): user data to pass to @callback
+ *
+ * Calls a method on this channel.
+ */
+void fl_method_channel_invoke_method(FlMethodChannel* channel,
+                                     const gchar* method,
+                                     FlValue* args,
+                                     GCancellable* cancellable,
+                                     GAsyncReadyCallback callback,
+                                     gpointer user_data);
+
+/**
+ * fl_method_channel_invoke_method_finish:
+ * @channel: a #FlMethodChannel
+ * @result:  #GAsyncResult
+ * @error: (allow-none): #GError location to store the error occurring, or %NULL
+ * to ignore.
+ *
+ * Completes request started with fl_method_channel_invoke_method().
+ *
+ * Returns: (transfer full): an #FlMethodResponse or %NULL on error.
+ */
+FlMethodResponse* fl_method_channel_invoke_method_finish(
+    FlMethodChannel* channel,
+    GAsyncResult* result,
+    GError** error);
+
+/**
+ * fl_method_channel_respond_success:
+ * @channel: a #FlMethodChannel
+ * @response_handle: handle that was provided in a
+ * #FlMethodChannelMethodCallHandler
+ * @result: (allow-none): value to respond with, must match what the
+ * #FlMethodCodec supports
+ * @error: (allow-none): #GError location to store the error occurring, or %NULL
+ * to ignore
+ *
+ * Responds to a method call.
+ *
+ * Returns: %TRUE on success.
+ */
+gboolean fl_method_channel_respond_success(
```

Comment:
Given that there is now a method response object in the API, can't these collapse to a single respond message that takes a response object?

Reply:
We could, but it's a bit of a pain to use in GObject, i.e.

```
g_autoptr(FlMethodSuccessResponse) response = fl_method_success_response_new(result);
fl_method_channel_respond(channel, response_handle, response, error);
```
vs
```
fl_method_channel_respond_success(channel, response_handle, result, error);
```

I considered adding `fl_method_channel_respond` and treating `fl_method_channel_respond_success`, `fl_method_channel_respond_error`, fl_method_channel_respond_not_implemented` as helper functions but in practise I don't think anyone would ever use the former.

Reply:
> in practise I don't think anyone would ever use the former.

In an isolated code snippet maybe not, but there are several patterns that the former allows (pseudocode, obviously):

```
response = nullptr;
if (method == "a") {
  [some logic]
  response = success
} else if (method == "b") {
  [some logic]
  response = success
} else {
  response = not implemented
}
send(response)
```

While you can do inline sends instead, this can be a useful structure to make sure you are always sending the response exactly once.

```
if (method == "a") {
  send(do_a_bunch_of_complex_stuff_that_may_succeed_or_have_errors_and_returns_a_response())
} else if (method == "b") {
  send(another_method_like_the_one_above())
} else {
  send(not implemented)
}
```

Without a generic response-send, you have to pass multiple bits of state into each sub-method, and they become more complicated conceptually because they have a side-effect of sending a response instead of returning a response.

I think the addition of one line in some cases is a small price to pay for the added flexibility. I've used both of the patterns above in writing plugins and system channel implementations.